File: src/leetcode/Javascript/202.Happy-Number.js
/*
* Write an algorithm to determine if a number n is happy.
*/
/**
 * @param {number} num
 * @return {boolean}
 */
const isHappy = (num) => {
  // Create a set/hashmap to store previous numbers that we've seen
  let seen = new Set();
  
  // Keep looking until we reach 1 or we've seen the same number before 
  while (num !== 1 && !seen.has(num)) {
    seen.add(num);
    
    //Replace the number with the sum of the squares of its digits 
    let nextNum = 0; 
    //convert number to string and split each digit 
    let digits =num.toString().split(''); 
    for(let i = 0; i < digits.length; i++){
      nextNum += parseInt(digits[i]) ** 2;
    }
    num = nextNum;
  }
  //return whether or not we reached 1
  return num === 1;
}

console.log(isHappy(19));  //Output: true
console.log(isHappy(2));  //Output: false
console.log(isHappy(22));  //Output: false

/*
* Time complexity - O(k) where k is the number is not a happy number. 
In the worst case, this could be an infiniite loop, but for most numbers it will be a relatively small constant.
*/

File: src/leetcode/python/202.Happy_Number.py
'''
Write an algorithm to determine if a number n is happy.

A happy number is a number defined by the following process:

Starting with any positive integer, replace the number by the sum of the squares of its digits.
Repeat the process until the number equals 1 (where it will stay), or it loops endlessly in a cycle which does not include 1.
Those numbers for which this process ends in 1 are happy.
Return true if n is a happy number, and false if not.
'''


class Solution:
    #Sln1-> Time Complexity O(k) Space Complexity O(m) 
    # Note that K represents the number of iterations while m the number of unique numbers encountered!!
    def isHappy(self, n: int)-> bool:
        seen = set()
        while n not in seen:
            seen.add(n)
            n = self.sumOfSquares(n)

            if n == 1:
                return True
        return False

    #helper function to get the sum of the squares
    def sumOfSquares(self, n: int)-> int:
        sum = 0
        while n:
            #gets the last digit
            digit = n % 10
            #squares it
            digit = digit ** 2
            #adds it to the sum
            sum += digit
            #removes the last digit by flooring the number
            n = n // 10
        return sum

    #Sln2->Time Complexity O(k) ->Space Complexity O(m) 
    def isHappy2(self, n: int)-> bool:
        seen = set()
        while n not in seen:
            seen.add(n)
            n = sum(int(d)**2 for d in str(n))
        return n == 1



s = Solution()
print(s.isHappy(19)) # Output-> True
print(s.isHappy(2)) #Output -> False

